Don't leak sysroot crates through dependencies

bjorn3 opened the pull request:
Previously if a dependency of the current crate depended on a sysroot crate, then `extern crate` would in the current crate would pick the first loaded version of said sysroot crate even in case of an ambiguity. This is surprising and brittle. For `-Ldependency=` we already blocked this since rust-lang/rust#110229, but the fix didn't account for sysroot crates.

Should fix https://github.com/rust-lang/rust/issues/147966

bjorn3 committed:
Don't leak sysroot crates through dependencies

Previously if a dependency of the current crate depended on a sysroot
crate, then extern crate would in the current crate would pick the first
loaded version of said sysroot crate even in case of an ambiguity. This
is surprising and brittle. For -Ldependency= we already blocked this,
but the fix didn't account for sysroot crates.

bjorn3 committed:
Remove PathKind from CrateSource

It is never used anywhere anymore now that existing_matches doesn't use
it anymore. And there is no good reason for any other code to use it in
the future either.

bjorn3 committed:
Fix rustc_hash ambiguity in miri

There are two rustc_hash crates in the sysroot when building miri and
there is no way to disambiguate between the two. Instead use the
re-exports of it in the rustc_data_structures crate.

bjorn3 committed:
Avoid multiple copies of proc macros ending up in the sysroot

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19667923867/job/56329347434) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56329347434) [(plain)](https://github.com/rust-lang/rust/commit/8d439c2c1ac1e0c3cf47978ef5694133b589d4b9/checks/56329347434/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_helper test:false 0.722
error: the borrowed expression implements the required traits
    --> src/bootstrap/src/core/build_steps/compile.rs:2430:44
     |
2430 |         builder.copy_link(&path, &dst.join(&filename), FileType::Regular);
     |                                            ^^^^^^^^^ help: change this to: `filename`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
     = note: `-D clippy::needless-borrows-for-generic-args` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(clippy::needless_borrows_for_generic_args)]`

```

</details>

petrochenkov commented on compiler/rustc_metadata/src/locator.rs:
```diff
@@ -798,21 +798,21 @@ impl<'a> CrateLocator<'a> {
             };
             if file.starts_with("lib") {
                 if file.ends_with(".rlib") {
-                    rlibs.insert(loc_canon.clone(), PathKind::ExternFlag);
```
Looks like `PathKind::ExternFlag` is no longer used anywhere.

bjorn3 commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -551,37 +550,19 @@ impl CStore {
-            }
+            // While the crate name matched, no --extern crate_name=path matched. It is possible
+            // that we have already loaded the target crate, but if that happens CStore::load will
+            // indicate so and we gracefully handle this, just potentially wasting a bit of time.
```
> If that's the case, then why do the new ambiguity errors appear?

Because the old `existing_match` accidentally resolved sysroot ambiguities in favor of the first loaded crate with the given name. The point of this PR is to remove this accidental ambiguity resolution.

> So technically we could eliminate existing_match entirely, and it wouldn't change the logic, only potentially introduce perf regressions?

If `existing_match` is bug free, then yes it should merely be a perf optimization.

petrochenkov replied:
> > That logic is also able to hide legitimate ambiguity errors - [#147966 (comment)](https://github.com/rust-lang/rust/issues/147966#issuecomment-3444198198).
> 
> Ignoring speculatively loaded crates in `existing_match` when no crate hash is given would work I presume?

I don't know :)
Need to investigate.

bjorn3 committed:
Add a test for not resolving ambiguity for multiple --extern with same name

bjorn3 committed:
Only check for already loaded indirect dependencies in existing_match

For direct dependencies it is a lot harder to not accidentally resolve
an ambiguity through an indirect dependency edge and as it turns out
this check was actually more expensive than the work it skipped. For
indirect dependencies existing_match is still necessary due to an assert
in load, but at the same time it also can't accidentally resolve an
ambiguity given that we already know exactly which crate to look for
based on the crate hash.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (c77897e64048c8406614d7cc2520bd22084bf335): [comparison URL](https://perf.rust-lang.org/compare.html?start=7b9905edb4df3aeaabd0d6cd0d4c09b183d7d965&end=c77897e64048c8406614d7cc2520bd22084bf335&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=7b9905edb4df3aeaabd0d6cd0d4c09b183d7d965&end=c77897e64048c8406614d7cc2520bd22084bf335&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.1%  | [0.1%, 0.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-2.7%, -0.4%] | 9     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7b9905edb4df3aeaabd0d6cd0d4c09b183d7d965&end=c77897e64048c8406614d7cc2520bd22084bf335&stat=max-rss)
<details>
<summary>Results (primary -0.3%, secondary -5.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.5%  | [1.5%, 1.5%]    | 1     |
| Regressions ❌ <br /> (secondary)  | 4.8%  | [4.8%, 4.8%]    | 1     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%]  | 1     |
| Improvements ✅ <br /> (secondary) | -7.9% | [-11.5%, -0.6%] | 4     |
| All ❌✅ (primary)                 | -0.3% | [-2.1%, 1.5%]   | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7b9905edb4df3aeaabd0d6cd0d4c09b183d7d965&end=c77897e64048c8406614d7cc2520bd22084bf335&stat=cycles:u)
<details>
<summary>Results (primary -2.0%, secondary 3.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.6%  | [2.6%, 4.6%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-2.1%, -1.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.0% | [-2.1%, -1.9%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=7b9905edb4df3aeaabd0d6cd0d4c09b183d7d965&end=c77897e64048c8406614d7cc2520bd22084bf335&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 470.06s -> 469.991s (-0.01%)
**Artifact size**: 386.96 MiB -> 386.96 MiB (-0.00%)

bjorn3 commented:
That turns out to be much better than I expected. It is a small perf win rather than a perf hit as I was afraid of. I've now cleaned up `existing_match` and made it immediately return `None` when no crate hash is given rather than still iterating over all crates, which should improve perf even more.

@rustbot ready

bjorn3 commented:
Oops, pushed to the wrong branch. Force pushed it back to the approved state.

@bors r=petrochenkov

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/19781441011/job/56682858493)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/19781441011/job/56682858493","builder_name":"checks-actions"} -->

bjorn3 commented:
@bors try jobs=dist-x86_64-msvc

bjorn3 commented:
@rustbot ready

rust-log-analyzer commented:
The job **`dist-powerpc64le-linux-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19860669540/job/56909757382) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56909757382) [(plain)](https://github.com/rust-lang/rust/commit/728aad8b696e14741324720b7d6a98109a5cbe24/checks/56909757382/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bjorn3 commented:
@bors try jobs=dist-powerpc64le-linux-gnu

jieyouxu commented:
Tagging this with compat relnotes, since it does break the crates that relied on this accidental behavior before.

@rustbot label: +relnotes
